### ISSUE TITLE
RSS_STAT_ARRAY compilation failure

### ISSUE THREAD

**Issue by albe19029** [falcosecurity/libs#2086]
When I try to build bpf driver I get next error:

[configure-bpf] Including /usr/src/scap-7.2.1+driver/bpf//configure/RSS_STAT_ARRAY/Makefile.inc
[configure-bpf] Build output for HAS_RSS_STAT_ARRAY:
[configure-bpf] make: Entering directory '/usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY' make -C /host/usr/src/linux-headers-4.15.0-213-generic M=$PWD clang  -I./arch/x86/include -I./arch/x86/include/generated  -I./include -I./arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h -Iubuntu/include  \     -D__KERNEL__   \       \      -D__KERNEL__ \  -D__BPF_TRACING__ \     -Wno-gnu-variable-sized-type-not-at-end \       -Wno-address-of-packed-member \         -fno-jump-tables \      -fno-stack-protector \  -Wno-tautological-compare \   -Wno-unknown-attributes \       -O2 -g -emit-llvm -c /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/test.c -o /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/test.ll In file included from /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/test.c:18: /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:85:17: error: use of undeclared identifier 'TASK_COMM_LEN'    85 |         char prev_comm[TASK_COMM_LEN];       |                        ^ /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:89:17: error: use of undeclared identifier 'TASK_COMM_LEN'    89 |         char next_comm[TASK_COMM_LEN];       |                        ^ /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:98:19: error: use of undeclared identifier 'TASK_COMM_LEN'    98 |         char parent_comm[TASK_COMM_LEN];       |                          ^ /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:100:18: error: use of undeclared identifier 'TASK_COMM_LEN'   100 |         char child_comm[TASK_COMM_LEN];       |                         ^ /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/test.c:27:20: error: subscripted value is not an array, pointer, or vector    27 |         val = mm->rss_stat[0].count;       |               ~~~~~~~~~~~~^~ 5 errors generated. make[2]: *** [/usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/Makefile:33: /usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY/test.o] Error 1 make[1]: *** [Makefile:1596: _module_/usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY] Error 2 make: *** [Makefile:26: all] Error 2 make: Leaving directory '/usr/src/scap-7.2.1+driver/bpf/configure/RSS_STAT_ARRAY'

This lead to invalid logic of HAS_RSS_STAT_ARRAY check, and rss_stat read invalid values for kernels > 6.2.

From what I can see TASK_COMM_LEN only defined for modern_bpf. Can this problem be fixed somehow?

**Comment by albe19029** [comment]
cat build.log
make: Entering directory '/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY'
make -C /host/usr/src/linux-headers-4.15.0-213-generic M=$PWD
make[1]: Entering directory '/host/usr/src/linux-headers-4.15.0-213-generic'
clang  -I./arch/x86/include -I./arch/x86/include/generated  -I./include -I./arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h -Iubuntu/include  \
        -D__KERNEL__   \
         \
        -D__KERNEL__ \
        -D__BPF_TRACING__ \
        -Wno-gnu-variable-sized-type-not-at-end \
        -Wno-address-of-packed-member \
        -fno-jump-tables \
        -fno-stack-protector \
        -Wno-tautological-compare \
        -Wno-unknown-attributes \
        -O2 -g -emit-llvm -c /usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/test.c -o /usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/test.ll
In file included from /usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/test.c:18:
/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:85:17: error: use of undeclared identifier 'TASK_COMM_LEN'
   85 |         char prev_comm[TASK_COMM_LEN];
      |                        ^
/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:89:17: error: use of undeclared identifier 'TASK_COMM_LEN'
   89 |         char next_comm[TASK_COMM_LEN];
      |                        ^
/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:98:19: error: use of undeclared identifier 'TASK_COMM_LEN'
   98 |         char parent_comm[TASK_COMM_LEN];
      |                          ^
/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/../../types.h:100:18: error: use of undeclared identifier 'TASK_COMM_LEN'
  100 |         char child_comm[TASK_COMM_LEN];
      |                         ^
/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/test.c:27:20: error: subscripted value is not an array, pointer, or vector
   27 |         val = mm->rss_stat[0].count;
      |               ~~~~~~~~~~~~^~
5 errors generated.
make[2]: *** [/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/Makefile:33: /usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY/test.o] Error 1
make[1]: *** [Makefile:1596: _module_/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY] Error 2
make[1]: Leaving directory '/host/usr/src/linux-headers-4.15.0-213-generic'
make: *** [Makefile:26: all] Error 2
make: Leaving directory '/usr/src/scap-7.2.0+driver/bpf/configure/RSS_STAT_ARRAY'

**Comment by albe19029** [comment]
So to fix the problem we need include this:
#include <linux/sched.h>

to RSS_STAT_ARRAY\test.c?

**Comment by FedeDP** [comment]
It seems so :) mind to test and eventually open a PR to fix?

**Comment by albe19029** [comment]
Probably yes - I will test and create a patch, is it OK?

**Comment by albe19029** [comment]
I have tested ubunt22 (5.15.0-116) and ubuntu24 (6.8.0-45-generic).
From the output I see that ubuntu 22 has this error:
error: subscripted value is not an array, pointer, or vector    27 |         val = mm->rss_stat[0].count;

The one we expect, not TASK_COMM_LEN.

And on ubuntu24 test module compiled success and I see -DHAS_RSS_STAT_ARRAY option in compile params.

So probably linux/sched.h was added to linux/mm_types.h or other header between 4.15 and 5.15. But as we check 6.2 kernel, don't think it is critical.
So just close this, as on distros >= 5.15 everything is working.
